pkg/storage/metrics.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Nuts crypto
+ * Copyright (C) 2020. Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package storage
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/nuts-foundation/nuts-crypto/log"
+	errors2 "github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// in minutes
+const checkInterval = 15
+
+const label = "period"
+
+// ErrAlreadyStarted is returned when start is called on an already running monitor
+var ErrAlreadyStarted = errors.New("certificate monitor already started")
+
+// we register only one
+var metric *prometheus.GaugeVec
+var mutex sync.Mutex
+
+// CertificateMonitor represents a go procedure which monitors expiring certificates within a given period.
+type CertificateMonitor struct {
+	exit        chan bool
+	Period      time.Duration
+	PeriodLabel string
+	Storage     Storage
+}
+
+// Start the certificate monitor for checking expiring certificates between now and the configure period.
+func (cm *CertificateMonitor) Start() error {
+
+	err := cm.init()
+	if err != nil {
+		return err
+	}
+
+	if cm.exit != nil {
+		return ErrAlreadyStarted
+	}
+
+	cm.exit = make(chan bool)
+	go cm.loop()
+
+	return nil
+}
+
+func (cm *CertificateMonitor) init() error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var err error
+
+	if metric == nil {
+		metric = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{Name: "nuts_crypto_certificate_expiry"},
+			[]string{label},
+		)
+		err = prometheus.Register(metric)
+	}
+
+	return err
+}
+
+// Stop the certificate monitor
+func (cm *CertificateMonitor) Stop() {
+	cm.exit <- true
```

Comment:
I'm actually not entirely sure whether this adds something. Blocking goroutines (waiting for tickers to fire) don't block the process from exiting and since we don't wait here for `loop()` to end graceful shutdown becomes subject for racy conditions (if the process exits before the `exit` channel has been read and shut down, I'm not considering it a 'graceful' shutdown, even when that shutdown doesn't actually do anything other than returning from the `loop()` function).

We have this construct all over the place though...

Reply:
without it, it blocks (at least the tests)